### ISSUE TITLE
impErrorTrap macro does not catch exceptions on Unix

### ISSUE THREAD

**Issue by jkotas** [dotnet/coreclr#3058]
Repro:

```
using System;
unsafe class My {
    static void * x;
    static IntPtr y;
    static void Main() {
        y = (IntPtr)x;
    }
}
```

```
csc -o+ test.cs
crossgen -readytorun -Platform_Assemblies_Paths `pwd` test.exe
```

Actual result on Unix: `E_FAIL while compiling My.Main`
Expected result as on Windows: No errors

**Comment by jkotas** [comment]
The problem is that the impErrorTrap macro in the JIT is not catching expected internal exception. @janvorli Could you please take a look?

**Comment by janvorli** [comment]
The reason why the impJitErrorTrap doesn't process the exception is that the exception is not thrown as PAL_SEHException. It is thrown at ZapSig::EncodeMethod src/vm/zapsig.cpp:1167 using the ThrowHR function. And this function uses EX_THROW which in turn uses plain throw. In our case, it throws HRException.
The impJitErrorTrap installs an exception filter, but that works for PAL_SEHException only, so the filter is ignored.

**Comment by BruceForstall** [comment]
Should EX_THROW call RaiseException instead?

**Comment by jkotas** [comment]
EX_THROW throws a typed C++ exception. It is a macro wrapper for `throw new MyException()`. There are similar EX_XXX macros wrappers for catching exceptions as well - that allow work as one would as one would except: allow catching exceptions of particular type, etc.

RaiseException can only throw exceptions of fixed "type", so it would not be a straightforward conversion.

We should be working towards removing the "special" VM exception handling from the JIT, to loosen the coupling between the two. The JIT should internally use regular portable C++ exceptions. JIT should not have assumptions about what type of exceptions are thrown by the VM. When the JIT needs to catch VM exceptions (e.g. around inlining), it should call back into the VM to do so.

**Comment by janvorli** [comment]
Assigned to @pgavlin since according to my knowledge, he is currently at some change in the JIT/EE interface as well as some refactoring to allow for a standalone DLL.

**Comment by pgavlin** [comment]
After digging into the current interactions between the JIT and the EE w.r.t. exceptions, I agree that the solution @jkotas proposed is probably the best option. Ideally we either would not allow exceptions to flow over the JIT/EE interface, but the way that `ICorJitInfo::FilterException` and `ICorJitInfo::HandleException` are used today seems to imply that this would not be straightforward. Adding a new method to ICJI that allows the JIT to at least remain ignorant of the exception types (if not the fact that certain JIT/EE interface methods may throw) seems to be a reasonable enough compromise.

**Comment by pgavlin** [comment]
FWIW, the original repro only works on Windows because C++ EH is implemented using SEH and the JIT's `__EEfilter` accepts (almost) all SEH exceptions, which are then passed to the EE for handling. If C++ EH did not interoperate with SEH on Windows, this test would fail there as well.

In light of this, should we consider deriving all of our C++ exception types from PAL_SEHException on *nix? There may be other instances of this problem lurking elsewhere in the code (i.e. code that uses PAL_EXCEPT_FILTER to catch a C++ exception).

**Comment by janvorli** [comment]
@pgavlin I think it would not work - we have places where we explicitly catch only exceptions that are not the PAL_SEHException. That's why we have two different sets of macros, PAL_CPP_TRY / PAL_CPP_CATCH_EXCEPTION / ... that catch exceptions derived from the Exception object and then PAL_TRY / PAL_EXCEPT / ... that are intended to catch only the exceptions raised by RaiseException. Obviously they catch the other exceptions on Windows as you've said, but that I believe is just a side effect that's not intended to be used.
@jkotas can you please confirm that?

**Comment by pgavlin** [comment]
> we have places where we explicitly catch only exceptions that are not the PAL_SEHException

Interesting--can you give me an example?

> I believe is just a side effect that's not intended to be used.

Whether or not it's a side effect, the inability of a `PAL_EXCEPT_FILTER` to process a C++ exception is a behavioral change w.r.t. the Windows CLR. It might not be a problem in practice, but we should probably audit the various filters/dispositions/etc to be sure.

**Comment by janvorli** [comment]
Hmm, although we have places where we catch only exceptions derived from the Exception class, using the EX_CATCH_CPP_ONLY or PAL_CPP_CATCH_DERIVED, that would actually work even if we derived our C++ exceptions from the PAL_SEHException. 
But that would also mean that the C++ exceptions would now additionally contain the CONTEXT and EXCEPTION_RECORD and those would need to be populated at the throw time, since the exception filters get them as a parameter. That would mean that all throws of C++ exceptions would need to be wrapped in a macro or something like that that would fill in these records in the exception.
So the throwing of C++ exceptions would become considerably more expensive both memory and CPU wise (initializing and filling the whole context).
Based on my experience with the coreclr code base, I believe that the whole coreclr carefully handles managed (plain SEH / PAL_SEHException) and C++ exceptions as separate concepts and the jitter is the only one that blurs the difference in the __EEfilter. And I agree with @jkotas that the jitter should be agnostic of how VM implements various kinds of exceptions.

**Comment by pgavlin** [comment]
> And I agree with @jkotas that the jitter should be agnostic of how VM implements various kinds of exceptions.

Just to be clear, I'm on the same page here.

**Comment by jkotas** [comment]
> should we consider deriving all of our C++ exception types from PAL_SEHException 

That would create undesirable coupling between CoreCLR Unix PAL and JIT. @pgavlin I see that you have created a set of workitems for standalone JIT build. It may be worth it to have one for PAL (stop linking with CoreCLR Unix PAL).

Also, with this fixed - you can consider compiling the standalone JIT without `/EHa` on Windows to get the right portable behavior that does not mix SEH and hardware exceptions with C++ exceptions.

**Comment by pgavlin** [comment]
> > should we consider deriving all of our C++ exception types from PAL_SEHException
>
> That would create undesirable coupling between CoreCLR Unix PAL and JIT

The worry that I'm addressing with this suggestion isn't really JIT-related; the JIT just happens to be one manifestation of the problem. Even once we loosen the coupling between the JIT and the EE w.r.t. EH, the runtime may still contain instances of the issue. Imagine e.g. that the following code existed in the runtime:

```
void throwing_function()
{
    ThrowHR(SOME_HRESULT);
}

void caller()
{
    PAL_TRY
    {
        throwing_function()
    }
    PAL_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
    {
        printf("in handler");
    }
    PAL_TRY_END
}
```

On Windows, that code would print "in handler"; on Linux, it would not (the exception from `ThrowHR` would propagate out of `caller`).

There may not be any such code in the runtime, but if there is, it could be a bug waiting to happen.

**Comment by gkhanna79** [comment]
Please undo https://github.com/dotnet/coreclr/blob/master/clrdefinitions.cmake#L134 as part of your change.

**Comment by gkhanna79** [comment]
Please undo the change in clrdefinitions.cmake and validate on non-Windows platforms to confirm everything works.

**Comment by gkhanna79** [comment]
This is blocking https://github.com/dotnet/coreclr/issues/4531

**Comment by pgavlin** [comment]
Closed by #4813
